Refactor gac-run

carloshorn opened the pull request:
This PR should close #54 by introducing a the new module `l1c_factory`.
This factory is used by the command line script pygac-run, which now is also able to process directories and tar archives.

The critical point is the change in the `Reader.read` method, which now takes an open file object as input. This change could lead to backwards incompatibility if someone uses this method in a custom script. If you see any risk, I could use the `utils.file_opener` (where open files are passed through) again in the read method.

mraspaud commented on pygac/l1c_factory.py:
```diff
@@ -0,0 +1,90 @@
+                open_file = file_opener(filename)
+            else:
+                open_file = file_opener(fileobj)
+            with open_file as fdes:
```
This could be simplified to:
```python
with file_opener(fileobj or filename):
...
```

carloshorn replied:
Good catch!

mraspaud commented on pygac/l1c_factory.py:
```diff
@@ -0,0 +1,90 @@
+
+class L1cFactory(object):
+    """Factory for level 1c files."""
+    __builders = {
```
One underscore is enough

mraspaud commented on bin/pygac-run:
```diff
@@ -90,19 +74,62 @@ def str2scanline(string):
+                    fileobj = a.extractfile(filename)
+                    try:
+                        l1c_factory(filename, start_line, 
+                                    end_line, fileobj=fileobj)
```
This usage of the `l1c_factory` instance is confusing. In computer science, a factory is supposed to return new object. Here, the call is actually performing the full read write cycle instead.

From what I understand, you want to factorize the reading and writing into one single call, so the way this class is used, I would call it more like a `processor` for example.

carloshorn replied:
True, my first attempt was to return the builder instance, but since its only purpose is its call method, there is really no point in returning it.

mraspaud commented on pygac/l1c_factory.py:
```diff
@@ -0,0 +1,90 @@
+LOG = logging.getLogger(__name__)
+
+
+class L1cBuilder(object):
```
I don't see the point of having this as a separate class. Why not just add a eg `process` function to the `Reader` class ? This would also allow to factorize the `main_klm` and `main_pod` functions to just call the reader's `process` function.

carloshorn replied:
Things like the `Reader._prepare_channels` could have gone into it, but you are right, adding a process function to the `Reader` makes it simpler.

mraspaud commented on pygac/l1c_factory.py:
```diff
@@ -0,0 +1,90 @@
+        LOG.info("Processing took: %s", str(datetime.datetime.now() - tic))
+
+
+class L1cFactory(object):
```
I don't think this needs to be a class, it would be clearer if it was a function returning the reader to use on the given file. It would be useful not only for doing the entire processing as you do in `__call__` but also for external libs that just want to know what reader to use for the file.
Down the line, in gac_run, you can then just do
```python
reader = find_out_what_reader_can_read_this_file(filename)
...
reader.read_and_save(...)
```
which IMO makes it clearer than the `L1cFactory` usage.

carloshorn replied:
In which module would you put this function, or should it have its own one?

mraspaud replied:
Good question. I suppose `reader.py` would be a good candidate, it that doesn't create circular imports. Otherwise, do we already have a `utils` module ?

mraspaud commented on pygac/l1c_factory.py:
```diff
@@ -0,0 +1,90 @@
+        # which checks if this file belongs to the reader
+        # Then the factory loops over the reader classes to determine
+        # who can read it.
+        # At least, this desission does not belong into the run script!
```
I absolutely agree! Having a mechanism to determine if a given reader can use the file is the cleanest solution.
Maybe some inspiration can be taken from this: https://github.com/pytroll/satpy/blob/master/satpy/readers/avhrr_l1b_gaclac.py#L103-L124
Also, I have stumbled upon cases where the file had been renamed, and the filename itself couldn't be trusted. However, the header of the files always contain the original name of the file, to we could parse this (using trollsift) to get more information what file it is, pod vs klm, gac vs lac.

carloshorn replied:
I will have a look and give it a try

mraspaud replied:
Thanks ! But that can be done in another PR, this one is already quite big as it is now :)

mraspaud commented:
Also, in the future, please try to submit smaller changes at the time. This PR is quite large but could have been split into multiple smaller ones (ie one feature at the time). It would have made the review process less time consuming... ;)

carloshorn commented on pygac/utils.py:
```diff
@@ -18,13 +18,74 @@
+        file - path to file or file object
+    """
+    close = True
+    if is_gzip(file):
```
Good point, `gzip.open` does not raise an exception immediately, but the following lines should do the job:
```python
@contextmanager
def file_opener(file):
    """Open a file depending on the input.

    Args:
        file - path to file or file object
    """
    # open file if necessary
    if is_file_object(file):
        open_file = file
        close = False
    else:
        open_file = open(file, mode='rb')
        close = True
    # check if it is a gzip file
    try:
        file_object = gzip.open(open_file)
        file_object.read(1)
    except OSError:
        file_object = open_file
    finally:
        file_object.seek(0)
    # provide file_object with the context
    try:
        yield file_object
    finally:
        if close:
            file_object.close()
```

carloshorn commented on pygac/tests/test_reader.py:
```diff
@@ -44,17 +44,6 @@ def setUp(self, interpolator, *mocks):
         self.interpolator = interpolator
         self.reader = GACReader()
 
-    @mock.patch('pygac.reader.open', mock.mock_open(read_data='normal'))
```
Sure, I just wanted to wait for your reply, to see if you agree with the changes.

carloshorn commented on pygac/klm_reader.py:
```diff
@@ -574,12 +574,11 @@ class KLMReader(Reader):
     tsm_affected_intervals = TSM_AFFECTED_INTERVALS_KLM
 
-    def read(self, filename):
+    def read(self, fileobj):
```
I expected this change to be critical, but it's an easy step to fix it. The function `file_opener` has no effect on open files, i.e. it does not close them.

carloshorn commented on pygac/reader.py:
```diff
@@ -113,41 +116,83 @@ def __init__(self, interpolate_coords=True, adjust_clock_drift=True,
+    @property
+    def filename(self):
+        """Get the property 'filename'."""
+        return self.__filename
```
You are right... Damn copy pastes of examples ;-)

carloshorn commented on pygac/reader.py:
```diff
@@ -113,41 +116,83 @@ def __init__(self, interpolate_coords=True, adjust_clock_drift=True,
+    @property
+    def filename(self):
+        """Get the property 'filename'."""
+        return self.__filename
```
It is already initialized in the `__init__`. I added the filename to the argument list and therefore moved it upwards to the initialization of the other given arguments.
Since you are using the `Reader.read` method in satpy, I wanted to ask, if you keep a single reader instance and reuse it for may files, or if one reader instance represents one file.
In the former case, I would suggest to give the `Reader.read` the optional argument `fileobj` as seen in `file_opener` and keep setting the filename attribute in this method.

carloshorn replied:
Question: Should the filename attribute be initialized with the given filename, or would you prefer the variable `Reader.head['data_set_name']`?

mraspaud replied:
I believe using 'data_set_name' would be more robust, right ?

carloshorn commented on bin/pygac-run:
```diff
@@ -90,19 +74,62 @@ def str2scanline(string):
-    reader(args.filename, args.start_line, args.end_line)
+    if args.config:
+        os.environ['PYGAC_CONFIG_FILE'] = args.config
+    from pygac.l1c_factory import L1cFactory
```
In this case, I postponed the import in order to be able to set the environment variable first.

mraspaud replied:
I understand. But in the end I think we should use the environment variable only to initialise the value, ie pass a variable around instead of relying on the environment variable everywhere.

carloshorn commented on bin/pygac-run:
```diff
@@ -90,19 +74,62 @@ def str2scanline(string):
+                    try:
+                        l1c_factory(filename, start_line, 
+                                    end_line, fileobj=fileobj)
+                    except Exception as e:
```
Right...

carloshorn commented on bin/pygac-run:
```diff
@@ -90,19 +74,62 @@ def str2scanline(string):
+        if tarfile.is_tarfile(path):
+            # archive
+            logger.info('Open archive "%s"' % str(path))
+            with tarfile.open(path) as a:
```
Good idea, I would also introduce a decorator for the `l1c_factory` (or `l1c_processor`) call which implements the try except logging clause to avoid the code repetition.

carloshorn commented:
Sorry for that lengthy PR, but it's hard to stop once started ;-)

Thank you for taking the time.

carloshorn commented:
Hi @mraspaud, 
Sorry again for the long PR.
Now, it seems to work. I manually tested it, but I will of cause write the unit tests for the new stuff.
Here, I wanted to summarize the changes witch should have gone into separate PRs:

1. Added the possibility to pass the config file to the run-script (`pygac-run`)
2. Added the possibility to pass directories to the run-script (`pygac-run`)
3. Added the possibility to pass tar files to the run-script (`pygac-run`)
3.1. Required to pass filename and file object to reader (all `XxxReader.read` methods)
3.2. Offered to outsource and refactor the open method (`Reader._open` -> `utils.file_opener`)
3.3. Required to set `reader.filename` in `reader.read` method
3.4. Offered to implement `Reader.filename` as property with the setter taking care of the format
4. Added the `Reader.save` method to simplify the file processing interface (reader main)
4.1. Required to implement `Reader._prepare_channels` to pass the channels to `gac_io.save_gac` in the correct order
5. Added `l1c_processor` to move the reader selection from `pygac-run` to the internal library.
5.1. Required to write `l1c_processor.get_reader` to implement the reader selection
5.2. Offered to implement the alternative constructor `Reader.fromfile`
5.3. Offered to introduce `ReaderError` as exception during the search for the reader
5.4. Required to introduce `reader._validate_header` to check if this file corresponds to the reader
5.5. Offered to implement a decorator in `pygac-run` to log failed processing trials

Open questions/issues (not to be handled in this PR :-)):
1. Use the header attribute `data_set_name` to set the attribute `reader.filename`. This would again simplify the `reader.read` interface, because then we don't need the filename attribute if a file object is given, because all the information would be gathered from inside the file. 
2. `pygac` should not expect any environment variable (especially the config path). The environment variable should only serve as a fallback if nothing has been given. That means that this variable should not be set in the __init__.py. At least no error log if everything is fine.

Many thanks again for your time for the code review!

mraspaud commented:
I'll check in more detail when I'm back from holiday on friday. Keep up the good work in the meanwhile !

carloshorn commented:
Okay, now the tests work. However, the feature of processing tar archives will not be available in python 2.7, because gzip.open cannot handle file objects as arguments. Since python 2 reached its end of life this year, I think that it is not worth to find another way to make it work.

mraspaud commented on pygac/lac_reader.py:
```diff
@@ -48,11 +48,10 @@ def _validate_header(self):
         )
-        if transfer_mode == 'GHRR':
-            raise ReaderError('Wrong transfer mode "%s"!' % transfer_mode)
+        if transfer_mode != 'LHRR':
```
IIRC it can also be 'HRPT'

carloshorn replied:
From the [KLM guide](https://www1.ncdc.noaa.gov/pub/data/satellite/publications/podguides/N-15%20thru%20N-19/pdf/2.5%20Section%208.0%20NOAA%20Level%201B%20Database.pdf) I got the following information:
```
Four characters identifying the data type and transmission method.
Recognized character groups are:
HRPT= HRPT (direct readout full resolution AVHRR)
GHRR= GAC (recorded reduced resolution AVHRR)
LHRR= LAC (recorded HRPT AVHRR)
FRAC=FRAC (recorded full resolution, full orbit data from Metop)
HIRX= HIRS/3 or HIRS/4 data set derived from GAC embedded TIP
AMAX= AMSU-A data set derived from GAC embedded TIP
AMBX= AMSU-B data set derived from GAC embedded TIP
MHSX=MHS data set derived from AIP
MHSS=MHS data derived from stored AIP
MHSH=MHS data derived from HRPT
HIRS= HIRS/3 or HIRS/4 data set derived from stored TIP
AMAS= AMSU-A data set derived from stored TIP
AMBS= AMSU-B data set derived from stored TIP
```
Therefore, I thought that GAC only allows GHRR, but please let me know what else can be recognized as GAC-file. Same holds for the other `XxxReader._validate` methods, please have a look at them to check if the selection rule is set correctly.

mraspaud replied:
We tried with HRPT and LHRR and both worked with the LAC reader. FRAC should work too, but I think @helgaweb didn't succeed reading it.

helgaweb replied:
Never tried to read FRAC, however, FRAC can be 1km or 4km.
'AVHRR FRAC data, the user may choose 1 km or 4 km as their default, available in swaths typically covering roughly an orbits worth of data.' for Metop Global Data Services
So, which is the resolution we are talking about? Then it might be LHRR or GHRR?

mraspaud replied:
I was thinking 1km. I though the metop data you had was FRAC, sorry :)

mraspaud commented:
> Okay, now the tests work. However, the feature of processing tar archives will not be available in python 2.7, because gzip.open cannot handle file objects as arguments. Since python 2 reached its end of life this year, I think that it is not worth to find another way to make it work.

Yes I agree. Python 2 support will be removed from this package anyway as soon as it becomes a hinder.

mraspaud commented on pygac/l1c_processor.py:
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2014, 2019 Pygac Developers
```
```suggestion
# Copyright (c) 2014, 2020 Pygac Developers
```

mraspaud commented on pygac/l1c_processor.py:
```diff
@@ -0,0 +1,84 @@
+
+"""l1c files processor
+
+"""
```
Should be one line, starting with capital letter and ending with a dot.

mraspaud commented on pygac/l1c_processor.py:
```diff
@@ -0,0 +1,84 @@
+    # Move the Reader in front of _Readers. Chance is high that the
+    # next file is of the same kind.
+    _Readers.insert(0, _Readers.pop(index))
+    return reader
```
I'm happy this function got added, it will make it easier to use pygac from other libraries. However, if I read this correctly, the entire file is read and then the file pointer is put back to the beginning of the file.

I would like to avoid having to read the full file for this, because if we want to sort out files in a directory full of gac/lac files, this will might slow down the search.

Maybe `_validate_header` should be called here instead of `fromfile` ?

carloshorn replied:
Welcome back from vacation!
It appears like it will read the entire file, but it actually does only read the header. The header validation is performed in `reader.read` just after reading the header. Rewinding the file and re-reading the header with the next reader should not (much) affect the file system, because python uses an internal buffer for open files which should still contain the relevant bytes.

mraspaud replied:
Thanks !
I'm not following: `fromfile` calls `read`, in which I can see:
```python
                self.scans = np.frombuffer(
                    fd_.read(count*self.scanline_type.itemsize),
                    dtype=self.scanline_type,
                    count=count)
```
or am I missing something ?

mraspaud replied:
So if the header matches, the file is read entirely, right ?

carloshorn replied:
Yes right, if a file is called with the wrong reader, the `read` method would raise the `ReaderError` just after reading the header and return to the loop in `get_reader`. Only if the header passes the validation, it reads the scans. The important thing is that `fd_.read` is never called without an argument, so it never reads more than necessary before validation.

mraspaud replied:
Ok, so something needs to be fixed here: either the function name, or the what is happening inside. `get_reader` for me means I provide a file and get back a reader class I can instantiate (which would be really useful in eg satpy)...

carloshorn replied:
I see. You want a `get_Reader` function to get the class. I only head the instance in mind...
Should be easy to implement. I would split the `read` into two parts, which are `read_header` and `read_scans`. Maybe the `read_header` should be a class method which returns the ars/tbm_header and the header, or raises the `ReaderError` if the header does not fit.
Then we can use this class method in `get_Reader` to select the appropriate reader and return the class.

mraspaud replied:
That would be great !

carloshorn replied:
Now, the reader has the classmethod `XxxReader.can_read` which returns `True` if the reader understands the header otherwise `False`. This can be used for checking files against a given reader as done in `get_reader_class`. 
I kept the `reader.fromfile` method as it is, because to me it appears natural to have an alternative constructor which initializes the whole instance from the given input in analogy to  `datetime.datetime.fromisoformat` for example.

ninahakansson commented:
Hi Carlos, Nice work with the PR! I have a small test-file that I use to test our software level1c4pps. 
We use pygac to read data from within Satpy. My test file has fewer scan_lines than expected and only a part of the last line is included. It seem np.fromfile can handle this broken line but np.frombuffer can not. I am not sure if broken lines are likely to appear for the archived uncropped data.

carloshorn commented:
Hi @ninahakansson, thanks for the feedback. Yes, in case of a cropped file `np.frombuffer` will raise `ValueError: buffer is smaller than requested size`. It is of cause possible to implement the known behavior from `np.fromfile`, but I would not let it pass silently. There should be at least a log message, and I would propose to read as many scans as possible from the file and fill the rest of the array with dummy values. Otherwise there would be a discrepancy in *count_of_data_records* from the header and the actual array length. What do you think @mraspaud and @sfinkens?

sfinkens commented:
@carloshorn @ninahakansson What is the current behaviour? Does `np.fromfile` fill the last incomplete scanline with dummy values or does it remove the line?

ninahakansson commented:
Traditional behaviour np.fromfile (pygac <=1.3.0) is to removed the broken line. Pygac 1.3.1 can not read files with broken lines. Is there a possibility that count_of_data_records attribute is wrong? We might miss some data or we might add lots of nodata lines. If it can happen that count_of_data_records is wrong I think it is better to just read all the file and handle the discrepancy in count_of_data_records from the header and the actual array length later. A log message is a great idea!

carloshorn commented:
Hi @sfinkens, `np.fromfile(fileobj, dtype=mytype, count=n)` will read up to `n*mytype.itemsize` bytes from `fileobj`, but return a shorter array if there are less bytes to read (without any warning). This means that is silently removes the lines. I agree that the reader should be able to handle this situation, but the reader should raise at least a warning saying that it could not read `head["count_of_data_records"]` from file. 
My question is more about how to deal with the situation. Should the reader continue with a shorter array, or update the *count_of_data_records`, or should `reader.scans` have the expected length of *count_of_data_records*, but fill the missing lines with dummy values?

mraspaud commented:
@ninahakansson so you mean that `.fromfile` will read lines before and after the broken line, while `.frombuffer` will stop at the broken line ?

mraspaud commented:
@carloshorn I would vote for issuing a log.warning message and updating the `count_of_data_records`

carloshorn commented:
Hi @ninahakansson, `np.fromfile` will try to read `scantype` from `fileobj` until the end of file, but not more than `count=count_of_data_records`. Is there anything stored after the scan lines inside the file? If yes and if your cropped file only reduces the number of scans keeping trailing bytes, the reader will interpret these trailing bytes as `scantype`.

Hello @mraspaud, I would additionally raise a `warning.warn`, because the reader faces an inconsistent situation. Even if the reader knows how to handle it, the user (in an interactive shell e.g. in a Jupyter notebook) should know that the results may behave unexpectedly. (similar to numpy with divide by zero -> `RuntimeWarning: divide by zero encountered in true_divide`)

ninahakansson commented:
> @ninahakansson so you mean that `.fromfile` will read lines before and after the broken line, while `.frombuffer` will stop at the broken line ?

No the broken line is always in the end of the file.

mraspaud commented:
Ok, so to make things clear: we're talking about an incomplete scanline at the end of the file, not a corrupt file somewhere in the middle of the file (which wouldn't be detectable anyway).
With that in mind, it would be nice that we could read as much as possible (as `fromfile` allowed).

Regarding warn, I'm fine with it being issued too. However, a limitation iiuc is that a warning is only issued once, even if `warning.warn` is called multiple times right ?

ninahakansson commented:
@carloshorn: I was just thinking what happens if count_of_data_records is wrong? Like =-9, or =999999. I do not know if it ever happens to the GAC files though, so maybe pygac does not need to handle that.  And if it can be trusted reading at most count_of_data_records is fine.

carloshorn commented:
@ninahakansson, good point, I have not taken into account wrong header entries. I would be fine with reading as many scan line types as possible from file. I just need a short confirmation if all these bytes actually represent scan lines or if there is anything else stored afterwards inside these files.

@mraspaud, yes warnings only appear once per code execution, e.g. only once for a Jypyter cell. Therefore, I agree that logging should be done in any case.

mraspaud commented:
@carloshorn there is no trailer to the files, so everything until the end is scanlines

abhaydd commented:
Thank you all for the recent contributions and update. My experience is that the count_of_data_records is very reliable and, at the most, we could cross-check it with total number of scans (not the scanline numbers). The situation that the count_of_data_records contains invalid value is extremely rare. I haven't come across any such case so far. So please try to keep the code simple and clean, without complicating it too much. Thanks.

carloshorn commented on pygac/l1c_processor.py:
```diff
@@ -0,0 +1,84 @@
+
+LOG = logging.getLogger(__name__)
+
+_Readers = [GACKLMReader, LACKLMReader, GACPODReader, LACPODReader]
```
Now it is called `_reader_classes`.

carloshorn commented on pygac/pod_reader.py:
```diff
@@ -465,31 +493,26 @@ def get_tsm_pixels(self, channels):
         return get_tsm_idx(channels[:, :, 0], channels[:, :, 1],
                            channels[:, :, 3], channels[:, :, 4])
 
+    def _prepare_channels(self):
```
Okay, I took the long one.

carloshorn commented:
@ninahakansson, in the last commits, I restored the behavior known from `np.fromfile` for the scan lines. 
@sfinkens, I got your point about not implementing too many features into the command line script provided by the library and leave spacial requirements to the user, e.g. reading certain compressions, or archive formats like zip or tar... Now, that the readers can handle file objects and the library provides a class selection method, it should be easy to write a custom command line script.

The question is: what should the command line script cover?
At least, it should cover the case where the user wants to process a single file.

However, I don't know if this is the typical use case.
My user story is the following: I get my data from ECMWF. The files come in tarballs and are individually gzip compressed. As a user I want to process (read, calibrate and navigate) the data (all files) that I download from ECMWF as cluster job, so that I get the information about success or failure from the log file.

If many of your users share this story. It might be a good idea to have this feature inside the run script, or provide a specific run script for this story. Otherwise, I am absolutely fine with dropping it from the current script.

carloshorn commented on pygac/l1c_processor.py:
```diff
@@ -0,0 +1,84 @@
+    return reader
+
+
+def l1c_processor(filename, start_line, end_line, fileobj=None):
```
Now, the module is called `runner` and the function is called `process_file` as suggested.

carloshorn commented on bin/pygac-run:
```diff
@@ -90,19 +77,72 @@ def str2scanline(string):
+        try:
+            processor(filename, *args, **kwargs)
+        except Exception:
+            logger.exception('Could not porcess "%s"' % str(filename))
```
Now, the exceptions are not caught for single file execution.

sfinkens replied:
@carloshorn Regarding the `debug` option I was thinking of something like this (from the top of my head):

```python
def wrapper(filename, debug, *args, **kwargs):
    try:
        processor(filename, *args, **kwargs)
    except Exception:
        if debug:
            raise
        else:
            logger.exception(...)
```

This could then be called like `processor(filename, args.debug, start_line, end_line)`

sfinkens replied:
Oh, and there's a little typo in the log message: `porcess` vs `process`

carloshorn commented:
With the last commit the PR hopefully contains all points that were discussed. In summary:

1. It restores the behavior known from `np.fromfile` when reading the scan lines, but warns and logs it.
2. The script `pygac-run` raises the exceptions if called with single files.
3. The methods and functions got more self-explanatory names.
4. There is a function to get the reader class that can be used in other libraries.
5. Added tests.

I hope that I did not miss anything.

ninahakansson commented:
Hi @carloshorn, Thanks for making the updates! I have tested the updated PR and my test case is now fine.

abhaydd commented:
@carloshorn Would you please explain your point 2? Does it mean I can't call pygac-run on just one individual, l1b orbit? I often run pygac just case-by-case basis for testing. Thanks.

carloshorn commented:
Hi @abhaydd, your use case should be fine. It refers to a comment by @sfinkens:

> For example this transforms every exception into a log message. What if the user wants to catch it, e.g. via return code in a shell script

It only means that the script if you call it with a single file and an error occurs, e.g. you call it with the repository README.md or some other invalid file, will raise the error instead of logging it and terminate successfully. On the other hand, if you call the script with a directory and there is a broken file or there is no matching TLE or what ever, you don't want the script to terminate. It will log the exception, but continue processing the other files.

sfinkens commented:
Ok, since you guys seem to like the directory/tarfile browsing I'm fine with it 🙂 But then I think it would be really useful if `pygac-run` had a `debug` flag to make it stop on the first exception. And maybe switch on logging. What do you think?

sfinkens commented on pygac/runner.py:
```diff
@@ -0,0 +1,87 @@
+    # creation of the instance.
+    with file_opener(fileobj or filename) as open_file:
+        Reader = get_reader_class(filename, fileobj=open_file)
+        reader = Reader.fromfile(filename, fileobj=open_file)
```
I think the workflow would be more clear if `reader.read()` was called explicitly here:

```python
reader_cls = get_reader_class(...)
reader = reader_cls()
reader.read(filename, fileobj=open_file)
reader.save(start_line, end_line)
```
At least this is how it is used in satpy: https://github.com/pytroll/satpy/blob/master/satpy/readers/avhrr_l1b_gaclac.py#L130 . Or is there a reason why we need the separate `Reader.fromfile` classmethod? Maybe I missed that.

Note that we already have a class called `Reader` so a different name should be used here to avoid conflicts.

mraspaud replied:
I agree with @sfinkens , this would make it clearer.

carloshorn replied:
Okay

sfinkens commented on pygac/reader.py:
```diff
@@ -116,38 +127,181 @@ def __init__(self, interpolate_coords=True, adjust_clock_drift=True,
+            buffer (bytes, bytearray): buffer to read from
+            count (int): number of expected scanlines
+        """
+        line_count = len(buffer) // self.scanline_type.itemsize
```
Maybe add a comment here, that it is possible to have an incomplete scanline in the end of a file which will be removed by this method?

sfinkens commented on pygac/reader.py:
```diff
@@ -116,38 +127,181 @@ def __init__(self, interpolate_coords=True, adjust_clock_drift=True,
+    def _get_calibrated_channels_uniform_shape(self):
+        """Prepare the channels as input for gac_io.save_gac"""
+        channels = self.get_calibrated_channels()
+        assert channels.shape[-1] == 6
```
@mraspaud Is using assertions in the code a good practice or should this be checked by the unit tests instead?

mraspaud replied:
I don't have anything agains the assertion here, but yes the unit tests should test this case.

carloshorn replied:
I would like to keep the assert at this location, because the Reader class implements the default behavior which assumes this condition, but of cause this does not replace a unit test.

abhaydd commented:
@carloshorn Thank you very much for the explanation.

carloshorn commented:
Our friend @stickler-ci has a problem with the FileNotFoundError...

carloshorn commented:
I know that this PR is already very huge, but after touching `__init__.py`... 
I would like to remove any code implementation from `__init__.py` (to avoid circular dependencies) and use it exclusively for exposing things like `__version__`, getter and setter of configuration and maybe the `get_reader` method.
How about moving `get_absolute_azimuth_angle_diff`, `centered_modulus` and `calculate_sun_earth_distance_correction` into `pygac.utils` and move the config logic into a class in a dedicated module, e.g. `pygac.configuration` and only keeping an instance of it in `__init__.py`?

mraspaud commented:
@carloshorn go ahead!

sfinkens commented on bin/pygac-run:
```diff
@@ -90,19 +80,93 @@ def str2scanline(string):
+                        help='pygac config file to be used')
+    parser.add_argument('-v', '--verbose', action='count', default=0,
+                        help='Explain what is being done, frequent occurrence'
+                             ' increases details')
```
Neat!

sfinkens commented:
@carloshorn With some files I'm getting an "Unable to read the file" error like this:
```
[ INFO pygac.runner 2020-03-02 16:28:02.203] Process file: input/NSS.GHRR.ND.D98348.S1242.E1422.B3939192.GC
[ DEBUG pygac.reader 2020-03-02 16:28:02.204] validate header
[ DEBUG pygac.reader 2020-03-02 16:28:02.204] GACKLMReader failed to read the file! ReaderError('Not able to decode the data set name!')
[ DEBUG pygac.reader 2020-03-02 16:28:02.204] validate header
[ DEBUG pygac.reader 2020-03-02 16:28:02.204] LACKLMReader failed to read the file! ReaderError('Not able to decode the data set name!')
[ DEBUG pygac.reader 2020-03-02 16:28:02.205] validate header
[ DEBUG pygac.reader 2020-03-02 16:28:02.205] GACPODReader failed to read the file! ReaderError('Not able to decode the data set name!')
[ DEBUG pygac.reader 2020-03-02 16:28:02.205] validate header
[ DEBUG pygac.reader 2020-03-02 16:28:02.205] LACPODReader failed to read the file! ReaderError('Not able to decode the data set name!')
Traceback (most recent call last):
  File "/cmsaf/nfshome/sfinkens/software/devel/pygac/bin/pygac-run", line 164, in <module>
    pygac.process_file(filename, start_line, end_line)
  File "/cmsaf/nfshome/sfinkens/software/devel/pygac/pygac/runner.py", line 84, in process_file
    reader_cls = get_reader_class(filename, fileobj=open_file)
  File "/cmsaf/nfshome/sfinkens/software/devel/pygac/pygac/runner.py", line 59, in get_reader_class
    raise ValueError('Unable to read the file "%s"' % filename)
ValueError: Unable to read the file "input/NSS.GHRR.ND.D98348.S1242.E1422.B3939192.GC"
```

I've uploaded them here: https://public.cmsaf.dwd.de/data/sfinkens/pygac_l1c_factory/

sfinkens commented:
Output of other files is identical to https://github.com/pytroll/pygac/commit/6849db7171d983a798be2881303e84e611e170ec

carloshorn commented:
> @carloshorn With some files I'm getting an "Unable to read the file" error like this:
> [...]
> I've uploaded them here: https://public.cmsaf.dwd.de/data/sfinkens/pygac_l1c_factory/

Hi @sfinkens,
I found the bug. In the case of the above files, the `data_set_name` in the header of these files was not empty, but filled with an invalid name. Therefore, these headers were rejected by the validation. Now, I check before validation if the header needs to correct the `data_set_name` and if yes, it tries to use the filename.
I locally tested the header extraction and it seems to work for me. I did not run a full test yet, because I did not download the TLEs for these files.
Could you please rerun your test? Thanks in advance

sfinkens commented:
@carloshorn Errors are gone, but now some files are 1 scanline longer than before:

```
(pygac) sfinkens@oflks230:/cmsaf/nfshome/sfinkens/software/devel/pygac/testbench/refactor> h5ls output_ref/ECC_GAC_avhrr_noaa10_99999_19861227T0013030Z_19861227T0206560Z.h5/image1 
data                     Dataset {13666, 409}
how                      Group
what                     Group
(pygac) sfinkens@oflks230:/cmsaf/nfshome/sfinkens/software/devel/pygac/testbench/refactor> h5ls output/ECC_GAC_avhrr_noaa10_99999_19861227T0013030Z_19861227T0206565Z.h5/image1
data                     Dataset {13667, 409}
how                      Group
what                     Group
(pygac) sfinkens@oflks230:/cmsaf/nfshome/sfinkens/software/devel/pygac/testbench/refactor>
```

sfinkens commented:
I've uploaded the TLE's as well: https://public.cmsaf.dwd.de/data/sfinkens/pygac_tles/

carloshorn commented:
Hi @sfinkens, 
Thanks for providing the test data. I know where the difference comes from. In the former version, in `PODReader.read`, where the scan lines are read from file, it passes the `number_of_scans` from the header to np.fromfile's count argument. Therefore, it only reads maximum `number_of_scans` scan lines. The current implementation reads until the end of file which in the above case contained another scan line.
I could return to the old behavior with a simple `max` condition in `Reader._read_scanlines` if wanted. 
Another possibility would be to read all possible scanlines and handle `len(reader.scans) != number_of_scans` later.

@ninahakansson already mentioned this:
> @carloshorn: I was just thinking what happens if count_of_data_records is wrong? Like =-9, or =999999. I do not know if it ever happens to the GAC files though, so maybe pygac does not need to handle that. And if it can be trusted reading at most count_of_data_records is fine.

and @abhaydd replied
> Thank you all for the recent contributions and update. My experience is that the count_of_data_records is very reliable and, at the most, we could cross-check it with total number of scans (not the scanline numbers). The situation that the count_of_data_records contains invalid value is extremely rare. I haven't come across any such case so far. So please try to keep the code simple and clean, without complicating it too much. Thanks.

Note that `count_of_data_records` is the klm header key for the pod `number_of_scans`.

What would you prefer?

mraspaud commented:
I would rather read everything and warn about any extra scanlines found.

carloshorn commented:
Hi @mraspaud and @sfinkens, 
I think we are close to finish. Now, its time for the final clean-up. One thing that I remember is updating the year in the Copyright line and check the author(s) list.
Furthermore, I would go through the log messages and see if the information provided for the given log level is appropriate.
Any other ideas, or tests that should be added?

sfinkens commented:
@carloshorn I will run my tests again, but removing the last scanline in the comparisons

mraspaud commented:
I'll try to have a look a bit later.

sfinkens commented:
@carloshorn Now I see small differences for a couple of orbits, always in the end:

https://public.cmsaf.dwd.de/data/sfinkens/pygac_l1c_factory/pics/

carloshorn commented:
@sfinkens, are the files which show the differences the same which had more scan lines? If yes, I guess it is related to some interpolation or smoothing kernel which now finds more neighbors. 
I will make a test with the `min` condition in `reader._read_scanlines` and see if my guess holds.

sfinkens commented:
@carloshorn No, [this one](https://public.cmsaf.dwd.de/data/sfinkens/pygac_l1c_factory/pics/ECC_GAC_avhrr_noaa14_99999_20011201T0618407Z_20011201T0812227Z.h5_image4.png) for example has the same number of scanlines in both cases

carloshorn commented:
Good news, even if the output image has identical number of scanlines, it is one of the cases with more scanlines in the input file. I guess they are masked out, but still used for interpolation.
As confirmation of my above guess, I did three runs with the file `NSS.GHRR.NG.D86361.S0013.E0206.B0142628.GC`, and I reproduce the old version's result if I include the `min` statement in `Reader._read_scanlines`:

![image](https://user-images.githubusercontent.com/27016093/76089865-c3b20100-5fba-11ea-83e0-ada82140d0f9.png)

sfinkens commented:
Nice catch @carloshorn ! I guess we don't want to use scanlines for interpolation which don't show up in the final output? What is actually interpolated here @mraspaud ?

mraspaud commented:
IIRC we interpolate the lons and lats in case of clock offset, so unless the lons and lats are erraneous, I would say we should use them. When the scanlines are masked, it's because the radiances are bad, right ?

sfinkens commented:
Scanlines are masked based on the quality flags, e.g. https://github.com/pytroll/pygac/blob/826941fae7945d9c24b8c1cda629333bfc8076d1/pygac/klm_reader.py#L698

But I don't know what they mean...

carloshorn commented:
The [pod](https://www1.ncdc.noaa.gov/pub/data/satellite/publications/podguides/TIROS-N%20thru%20N-14/pdf/NCDCPOD3.pdf) guide **Table 3.1.2.1-2. Format of quality indicators** and [klm](https://www1.ncdc.noaa.gov/pub/data/satellite/publications/podguides/N-15%20thru%20N-19/pdf/0.0%20NOAA%20KLM%20Users%20Guide.pdf) guide **Table 8.3.1.3.2.1-1.** and following tables give the details about each bit. These tables should be mentioned in the doc string of `_get_corrupt_mask`. Furthermore, the bit shift operations in this method(s) are very cryptic to read. Since this method is far from being the bottleneck, using `np.binary_repr` to split the array into its bits and give them the associated names would make the it more readable end self-explanatory.

sfinkens commented:
Nice, thanks for clarifying and immediately fixing it! If the `NO_EARTH_LOCATION` flag is set, we probably shouldn't use the scanline for lat/lon interpolation. Of course we could start splitting the quality flags, but since we have quite a lot of changes in this PR already, I'd vote for restoring the original behaviour.

carloshorn commented:
> Nice, thanks for clarifying and immediately fixing it! If the `NO_EARTH_LOCATION` flag is set, we probably shouldn't use the scanline for lat/lon interpolation.

I had a look into the interpolator and would recommend to add an optional argument for a mask to the interpolation functions. This argument should be passed to the `lat_lot_interpolator` and mask the missing lines. It could look like this:

```python
def lat_lon_interpolator(lons_subset, lats_subset, cols_subset, cols_full, mask=None):
    """Interpolate lat-lon values in the AVHRR data."""
    lines = lats_subset.shape[0]
    rows_subset = np.arange(lines)
    if mask is not None:
        rows_subset = rows_subset[mask]
    rows_full = np.arange(lines)

    along_track_order = 1
    cross_track_order = 3

    satint = gtp.SatelliteInterpolator((lons_subset, lats_subset),
                                       (rows_subset, cols_subset),
                                       (rows_full, cols_full),
                                       along_track_order,
                                       cross_track_order)

    return satint.interpolate()
```
What do you think? This time, I maybe wait before implementing...

> Of course we could start splitting the quality flags, but since we have quite a lot of changes in this PR already, I'd vote for restoring the original behaviour.

I will revert to 487660d9a7499f7386bae016cb32ada773e369b8 and keep the commits for another PR.

mraspaud commented:
Yes, can we take this in an other PR ?

mraspaud commented:
I'd like to stop the addition of features on this one for now, so that we can make a final test and merge if possible

carloshorn commented:
As announced earlier, I reverted the bit unpacking commits and updated the author lists.
@mraspaud could you specify which tests you would like to see to close this PR.
Once closed, we can handle the bit unpacking, and lon/lat mask for interpolation in separate PRs.

mraspaud commented:
@carloshorn ok, thanks.
@ninahakansson @sfinkens could you please check that this PR works for you one last time ?

ninahakansson commented:
@mraspaud PR works ok for my testcase.

sfinkens commented:
Testing now

sfinkens commented:
So, a couple of files are longer than before. Was the consensus to use these extra scanlines? Apart from that, I see some differences in two files
```
NSS.GHRR.NJ.D01335.S0618.E0812.B3567273.WI
NSS.GHRR.NJ.D01335.S1454.E1639.B3567778.GC
```
See: https://public.cmsaf.dwd.de/data/sfinkens/pygac_l1c_factory

mraspaud commented:
@sfinkens aren't these just extra lines ?

sfinkens commented:
@mraspaud I don't think so, both orbits have the same length and if you look at `image4` for example, they look similar in the end, but the difference is not zero (although it is small)

sfinkens commented:
Ah, but now I remember. This was because the extra lines are used for interpolation, right?

mraspaud commented:
Yes that's probably the case.

mraspaud commented:
So it looks like everyone is good with this, I'll be merging shortly. @carloshorn thank you very much for your sustained efforts and patience! That was an epic PR :)

sfinkens commented:
Yeah, big thumbs up @carloshorn!

carloshorn commented:
Nice, thank you all for the reviews.
